src/ast/ast.c

```diff
@@ -0,0 +1,328 @@
+        }
+    }
+    return clause_indices;
+}
+
+uint AST_GetClauseCount(const AST *ast, cypher_astnode_type_t clause_type) {
+    uint clause_count = cypher_ast_query_nclauses(ast->root);
+    unsigned int num_found = 0;
+    for (unsigned int i = 0; i < clause_count; i ++) {
+        const cypher_astnode_t *child = cypher_ast_query_get_clause(ast->root, i);
+        if (cypher_astnode_type(child) == clause_type) num_found ++;
+
+    }
+    return num_found;
+}
+
+uint AST_NumClauses(const AST *ast) {
+    return cypher_astnode_nchildren(ast->root);
+}
+
+// TODO should just use this format or the TopLevel format
+const cypher_astnode_t** AST_CollectReferencesInRange(const AST *ast, cypher_astnode_type_t type) {
+    const cypher_astnode_t **found = array_new(const cypher_astnode_t *, 0);
+
+    uint clause_count = cypher_ast_query_nclauses(ast->root);
+    for (uint i = 0; i < clause_count; i ++) {
+        const cypher_astnode_t *child = cypher_ast_query_get_clause(ast->root, i);
+        if (cypher_astnode_type(child) != type) continue;
+
+        found = array_append(found, child);
+    }
+
+    return found;
+}
+
+void _AST_CollectAliases(const char ***aliases, const cypher_astnode_t *entity) {
+    if (entity == NULL) return;
+
+    if (cypher_astnode_type(entity) == CYPHER_AST_IDENTIFIER) {
+        const char *identifier = cypher_ast_identifier_get_name(entity);
+        *aliases = array_append(*aliases, identifier);
+        return;
+    }
+
+    uint nchildren = cypher_astnode_nchildren(entity);
+    for (uint i = 0; i < nchildren; i ++) {
+        _AST_CollectAliases(aliases, cypher_astnode_get_child(entity, i));
+    }
+}
+
+// Collect aliases from clauses that introduce entities (MATCH, MERGE, CREATE, UNWIND)
+const char** AST_CollectAliases(AST *ast) {
+    const char **aliases = array_new(const char*, 1);
+    uint clause_count = cypher_ast_query_nclauses(ast->root);
+    for (uint i = 0; i < clause_count; i ++) {
+        const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
+        cypher_astnode_type_t type = cypher_astnode_type(clause);
+        if (type == CYPHER_AST_MATCH  ||
+            type == CYPHER_AST_MERGE  ||
+            type == CYPHER_AST_CREATE ||
+            type == CYPHER_AST_UNWIND
+            ) {
+            _AST_CollectAliases(&aliases, clause);
+        }
+    }
+
+    return aliases;
+}
+
+const cypher_astnode_t* AST_GetBody(const cypher_parse_result_t *result) {
+    const cypher_astnode_t *statement = cypher_parse_result_get_root(result, 0);
+    assert(statement && cypher_astnode_type(statement) == CYPHER_AST_STATEMENT);
+
+    return cypher_ast_statement_get_body(statement);
+}
+
+AST* AST_Build(cypher_parse_result_t *parse_result) {
+    AST *ast = rm_malloc(sizeof(AST));
+    ast->entity_map = NULL;
+    ast->root = AST_GetBody(parse_result);
+    assert(ast->root);
+    if (cypher_astnode_type(ast->root) == CYPHER_AST_QUERY) AST_BuildEntityMap(ast);
```

Comment:
which values can we expect `cypher_astnode_type(ast->root)` to return at this point?

Reply:
Our only handled cases right now are CYPHER_AST_QUERY, CYPHER_AST_CREATE_NODE_PROP_INDEX, and CYPHER_AST_DROP_NODE_PROP_INDEX